File: src/repositories/user.repository.ts
import { Collection, ObjectId } from "mongodb";
import { getCoreDatabase } from "@/config/database";
import type { FIUser } from "@/types/user/fi_user";
import { logger } from "@/utils/logger";

export class UserRepository {
  private getCollection(): Collection<FIUser> {
    return getCoreDatabase().collection<FIUser>("users");
  }

  async create(user: FIUser): Promise<FIUser> {
    try {
      await this.getCollection().insertOne(user as any);
      logger.info("User created", { userId: user._id });
      return user;
    } catch (error) {
      logger.error("Failed to create user", error);
      throw error;
    }
  }

  async findById(_id: string): Promise<FIUser | null> {
    try {
      return await this.getCollection().findOne({ _id } as any);
    } catch (error) {
      logger.error("Failed to find user by ID", error);
      throw error;
    }
  }

  async findByPhoneNumber(phoneNumber: string): Promise<FIUser | null> {
    try {
      return await this.getCollection().findOne({ phoneNumber } as any);
    } catch (error) {
      logger.error("Failed to find user by phone number", error);
      throw error;
    }
  }

  async findByCompanyId(companyId: string): Promise<FIUser[]> {
    try {
      return await this.getCollection().find({ companyId } as any).toArray();
    } catch (error) {
      logger.error("Failed to find users by company ID", error);
      throw error;
    }
  }

  async update(id: string, updates: Partial<FIUser>): Promise<FIUser | null> {
    try {
      const result = await this.getCollection().findOneAndUpdate(
        { _id: id } as any,
        { $set: updates },
        { returnDocument: "after" }
      );
      return result as FIUser | null;
    } catch (error) {
      logger.error("Failed to update user", error);
      throw error;
    }
  }
}

File: fisoft/src/main/kotlin/repository/UserRepository.kt
package com.ilyasdemirkiran.repository

import com.ilyasdemirkiran.config.Database
import com.ilyasdemirkiran.types.FIUser
import com.ilyasdemirkiran.types.UserRole
import com.mongodb.client.model.Filters
import com.mongodb.client.model.FindOneAndUpdateOptions
import com.mongodb.client.model.ReturnDocument
import com.mongodb.client.model.Updates
import com.mongodb.kotlin.client.coroutine.MongoCollection
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.toList

class UserRepository {
    private fun getCollection(): MongoCollection<FIUser> {
        return Database.getCoreDatabase().getCollection<FIUser>("users")
    }

    suspend fun create(user: FIUser): FIUser {
        getCollection().insertOne(user)
        return user
    }

    suspend fun findById(id: String): FIUser? {
        return getCollection().find(Filters.eq("_id", id)).firstOrNull()
    }

    suspend fun findByPhoneNumber(phoneNumber: String): FIUser? {
        return getCollection().find(Filters.eq("phoneNumber", phoneNumber)).firstOrNull()
    }

    suspend fun findByCompanyId(companyId: String): List<FIUser> {
        return getCollection().find(Filters.eq("companyId", companyId))
            .toList()
            .filter { it.role != UserRole.sudo }
    }

    suspend fun update(id: String, name: String? = null, surname: String? = null, companyId: String? = null, role: UserRole? = null): FIUser? {
        val updates = mutableListOf<org.bson.conversions.Bson>()
        if (name != null) updates.add(Updates.set("name", name))
        if (surname != null) updates.add(Updates.set("surname", surname))
        if (companyId != null) updates.add(Updates.set("companyId", companyId))
        if (role != null) updates.add(Updates.set("role", role.name))

        if (updates.isEmpty()) return findById(id)

        return getCollection().findOneAndUpdate(
            Filters.eq("_id", id),
            Updates.combine(updates),
            FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER)
        )
    }

    suspend fun delete(id: String): Boolean {
        val res = getCollection().deleteOne(Filters.eq("_id", id))
        return res.deletedCount > 0
    }
}
